ee/codegen/src/generators/graph-attribute.ts

```diff
@@ -0,0 +1,409 @@
+type GraphEmpty = { type: "empty" };
+type GraphSet = { type: "set"; values: GraphMutableAst[] };
+type GraphNodeReference = {
+  type: "node_reference";
+  reference: BaseNodeContext<WorkflowDataNode>;
+};
+type GraphPortReference = {
+  type: "port_reference";
+  reference: PortContext;
+};
+type GraphRightShift = {
+  type: "right_shift";
+  lhs: GraphMutableAst;
+  rhs: GraphMutableAst;
+};
+type GraphMutableAst =
+  | GraphEmpty
+  | GraphSet
+  | GraphNodeReference
+  | GraphPortReference
+  | GraphRightShift;
+
+export declare namespace GraphAttribute {
+  interface Args {
+    entrypointNodeEdges: WorkflowEdge[];
+    entrypointNodeId: string;
+    edgesByPortId: Map<string, WorkflowEdge[]>;
+    workflowContext: WorkflowContext;
+  }
+}
+
+export class GraphAttribute extends AstNode {
+  private readonly workflowContext: WorkflowContext;
+  private readonly edgesByPortId: Map<string, WorkflowEdge[]>;
+  private readonly entrypointNodeEdges: WorkflowEdge[];
+  private readonly entrypointNodeId: string;
+  private readonly astNode: python.AstNode;
+
+  public constructor({
+    entrypointNodeEdges,
+    entrypointNodeId,
+    edgesByPortId,
+    workflowContext,
+  }: GraphAttribute.Args) {
+    super();
+    this.entrypointNodeEdges = entrypointNodeEdges;
+    this.entrypointNodeId = entrypointNodeId;
+    this.workflowContext = workflowContext;
+    this.edgesByPortId = edgesByPortId;
+
+    const graphMutableAst = this.generateGraphMutableAst();
+    this.astNode = this.getGraphAttributeAstNode(graphMutableAst);
+    this.inheritReferences(this.astNode);
+  }
+
+  /**
+   * Generates a mutable graph AST.
+   *
+   * The algorithm we implement is a Breadth-First Search (BFS) that traverses through
+   * the edges of the graph, starting from the entrypoint node.
+   *
+   * The core assumption made is that `graphMutableAst` is always a valid graph, and
+   * adding a single edge to it will always produce another valid graph.
+   */
+  private generateGraphMutableAst(): GraphMutableAst {
+    let graphMutableAst: GraphMutableAst = { type: "empty" };
+    const edgesQueue = [...this.entrypointNodeEdges];
+    const processedEdges = new Set<WorkflowEdge>();
+
+    while (edgesQueue.length > 0) {
+      const edge = edgesQueue.shift();
+      if (!edge) {
+        continue;
+      }
+
+      const sourceNode =
+        edge.sourceNodeId === this.entrypointNodeId
+          ? null
+          : this.workflowContext.getNodeContext(edge.sourceNodeId);
+
+      const targetNode = this.workflowContext.getNodeContext(edge.targetNodeId);
+      if (!targetNode) {
+        processedEdges.add(edge);
+        continue;
+      }
+
+      const isPlural = (mutableAst: GraphMutableAst): boolean => {
+        return (
+          mutableAst.type === "right_shift" ||
+          (mutableAst.type === "set" && mutableAst.values.every(isPlural))
+        );
+      };
+
+      const getAstSources = (
```

Comment:
Am planning to pull out these nested helpers in a followup PR. Wanted to first do a logicless migration